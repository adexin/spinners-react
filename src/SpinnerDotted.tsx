import React, { CSSProperties } from 'react';
import { defaultProps, SpinnerProps } from './helpers';
import { SpinnersProps, withSharedProps } from './withSharedProps';

import './SpinnerDotted.css';

const coords = [
  { x: 22, y: -20 },
  { x: 29, y: 0 },
  { x: 22, y: 20 },
  { x: 0, y: 30 },
  { x: -23, y: 20 },
  { x: -30, y: 0 },
  { x: -23, y: -20 },
  { x: 0, y: -30 },
];

export type SpinnerDottedProps = SpinnersProps & SpinnerProps;

export const Component = ({
  speed = defaultProps.speed,
  still = defaultProps.still,
  thickness = defaultProps.thickness,
  ...svgProps
}: SpinnerProps) => {
  const duration = 200 / speed;
  const generateCircleStyle = (i: number): CSSProperties => (!still
    ? { animation: `spinners-react-dotted-shrink ${duration}s cubic-bezier(0, 0.9, 0, 0.9) ${(duration / 20) * i}s infinite` }
    : {});
  const centerStyle: CSSProperties = !still
    ? {
      animation: `spinners-react-dotted-center ${duration}s ease-out infinite`,
      transformOrigin: 'center',
    }
    : { display: 'none' };

  return (
    <svg fill="none" viewBox="0 0 66 66" {...svgProps}>
      {coords.map((c, i) => (
        <circle
          key={`${c.x}-${c.y}`}
          cx="33"
          cy="33"
          fill="currentColor"
          r={3 * (thickness / 100)}
          style={{
            transform: `translate(${c.x}px, ${c.y}px)`,
            ...generateCircleStyle(i),
          }}
        />
      ))}
      <circle
        cx="33"
        cy="33"
        fill="currentColor"
        r={6 * (thickness / 100)}
        style={centerStyle}
      />
    </svg>
  );
};

export const SpinnerDotted = withSharedProps(Component);
