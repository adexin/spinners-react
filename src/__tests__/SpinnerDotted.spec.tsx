import { render } from '@testing-library/react';
import React from 'react';

import { SpinnerDotted } from '../SpinnerDotted';

describe('SpinnerDotted', () => {
  const color = 'red';

  it('matches the snapshot', () => {
    const { container } = render(
      <SpinnerDotted color="#fff" size={100} speed={10} thickness={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const { container } = render(<SpinnerDotted />);

    expect(container).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const { container } = render(<SpinnerDotted enabled={false} />);

    expect(container.firstChild).toBe(null);
  });

  it('renders still if specified', () => {
    const { container } = render(<SpinnerDotted still />);
    const circles = container.querySelectorAll('circle');
    const last = circles.length - 1;

    expect(circles[0].style.animation).toBe('');
    expect(circles[last].style.animation).toBe('');
    expect(circles[last]).toHaveStyle({ display: 'none' });
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const { container } = render(
      <SpinnerDotted
        color={color}
        size={size}
        speed={speed}
        thickness={thickness}
      />,
    );
    const circles = container.querySelectorAll('circle');
    const last = circles.length - 1;
    const animationDuration = `${200 / speed}`;

    expect(container.firstChild).toHaveStyle({ width: size });
    expect(circles[0]).toHaveAttribute('r', `${3 * (thickness / 100)}`);
    expect(circles[0].style.animation.includes(animationDuration)).toBe(true);
    expect(circles[last]).toHaveAttribute('r', `${6 * (thickness / 100)}`);
    expect(circles[last].style.animation.includes(animationDuration)).toBe(
      true,
    );
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const { container } = render(
      <SpinnerDotted className={className} color="green" style={{ color }} />,
    );

    expect(container.firstChild).toHaveClass(className);
    expect(container.firstChild).toHaveStyle({ color });
  });
});
