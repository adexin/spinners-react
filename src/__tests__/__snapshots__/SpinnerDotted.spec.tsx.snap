// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerDotted matches the snapshot 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(255, 255, 255); overflow: visible; width: 100px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(22px, -20px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 0s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(29px, 0px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 1s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(22px, 20px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 2s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(0px, 30px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 3s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(-23px, 20px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 4s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(-30px, 0px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 5s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(-23px, -20px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 6s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="1.5"
      style="transform: translate(0px, -30px); animation: spinners-react-dotted-shrink 20s cubic-bezier(0, 0.9, 0, 0.9) 7s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="animation: spinners-react-dotted-center 20s ease-out infinite; transform-origin: center;"
    />
  </svg>
</div>
`;

exports[`SpinnerDotted matches the snapshot with default props 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(56, 173, 72); overflow: visible; width: 50px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(22px, -20px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(29px, 0px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.1s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(22px, 20px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.2s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(0px, 30px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.30000000000000004s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(-23px, 20px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.4s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(-30px, 0px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.5s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(-23px, -20px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.6000000000000001s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="3"
      style="transform: translate(0px, -30px); animation: spinners-react-dotted-shrink 2s cubic-bezier(0, 0.9, 0, 0.9) 0.7000000000000001s infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="currentColor"
      r="6"
      style="animation: spinners-react-dotted-center 2s ease-out infinite; transform-origin: center;"
    />
  </svg>
</div>
`;
